### ISSUE TITLE
Add `calc_first_two_phase_deriv_splined` to tabular backend

### ISSUE THREAD

**Issue by JonWel** [CoolProp/CoolProp#870]
Needed for the rho smoothing abilities in ExternalMedia.
See https://github.com/CoolProp/ExternalMedia/issues/16

**Comment by ibell** [comment]
Ugh, that is not going to be fun.

**Comment by JonWel** [comment]
As tabular backend are often used for speed purpose, this may be interesting in an `ExternalMedia` point of view to have a single function call to compute "`d_rho_spline_dp__consth` & `d_rho_spline_dh__constp`" or "`d_rho_spline_dp__consth` & `d_rho_spline_dh__constp` & `rho_spline`" as they share code and theses two combinations are used together in `ExternalMedia` when spline option is activated.
Or maybe a caching system to reuse the already computed values for successive call of the function?
Of course, the time involved may not worth this compare to all the other calls made by `ExternalMedia`.

**Comment by ibell** [comment]
That's not a bad idea.  First let me implement the splined derivatives for
the tabular backends, then we can revisit this issue.  I realized that
implementing the splined derivatives should actually be quite simple (I
think). I can move the implementation from the HEOS backend to the base
class, and that might be all that I need to do.  To be determined.

**Comment by JonWel** [comment]
Yes, the important think is to have the functionality. The speed improvement is secondary, and I may be able to do it myself (once the other function is there, I don't feel knowing enough on the tabular backend yet to add myself this function). This is more to have it written somewhere so I can come back on it later.

**Comment by jowr** [comment]
Any news here? Did any of you work on this issue?

**Comment by ibell** [comment]
I have not, having no need for this in my work.  In principle the thing to
do is to push the changes up the inheritance tree, perhaps all the way up
to AbstractState, so that we can generalize this function for any backend
that wants this functionality.  That would be my proposal, but you are
welcome to do as you please.

**Comment by jowr** [comment]
I think my dirty hack works, but this [workaround](https://github.com/CoolProp/CoolProp/blob/ef0f5284010e6cb69973c582ddbadd42aa67bb23/src/Backends/Tabular/TabularBackends.cpp#L750) should be addressed before we can close this issue.
![870_test_two_phase_tabular](https://cloud.githubusercontent.com/assets/769593/20684026/91f6edec-b5ae-11e6-8e4e-f4ebdc460b1c.png)

**Comment by ibell** [comment]
Looks like your dirty hack does work, but I still think it would be better
to push this into the AbstractState so that we can do this for REFPROP
too.  Might be worth to do some timing to see how bad the speed penalty
would be if we created two instances at runtime when the function gets
called.

**Comment by jowr** [comment]
Yes I agree, a generalised solution would be nice. I, however, do see a challenge here since we need access to two additional instances of the AbstractState object to calculate the start point and the end point of the spline. 

For the tabular methods, a full copy of the state object would involve copying a lot of data in the memory due to the tabulated data. This might take longer than just calling the backend that was used to generate the tables with two times with the correct rho and T inputs as I have done now. For all other backends, such a copy/clone operation should be OK. 

One possible solution could be to introduce a function that duplicates the state object referencing the already loaded tabular data in memory by means of a pointer to the original state object's tables. Such a function could be called `virtual AbstractState clone_linked(void)` and should be protected to make sure that the cloned object cannot be used outside the scope of the original state object to avoid memory leaks. For all other backends, such a function would just copy the whole object. This function could be added to the AbstractState and only the tabular backend should overwrite it to handle the memory issue. 

I do not know if this is at all possible since returning an object from a function might implicitly lead to a copy operation. An alternative could be to implement the functions `virtual void push_state(void)` and `virtual void pop_state(void)` to save and load state points using an internal stack. This could be implemented by saving the two properties that allow us to update the state object as fast as possible (rho, T, phase? for HEOS and REFPROP; ??? for tabular).

**Comment by ibell** [comment]
A generalized solution might be best, but it is also true that tabular
backends load from file automatically, and do not re-generate the tables,
even though there is some overhead for unpacking the tables, unzipping the
data, etc.

I've already started implementing a get_copy() function for the cubic
backends, and perhaps we want to consider doing the same for all backends.

**Comment by ibell** [comment]
Closing as part of an issue-tracker cleanup — this issue has had no activity in over two years.

*To be transparent: this comment and the close action are being performed by Claude (Anthropic's AI coding assistant), running on behalf of @ibell. Ian reviewed and approved the bulk-close criteria; I am the one executing the action.*

If you would like this issue reconsidered, please feel free to **re-open it** — a fresh comment with current behaviour against the latest CoolProp release is the most useful way to revive the discussion.

— Claude (acting for @ibell)
